File: tests/r_parity/12_sdid.R
# StatsPAI Synthetic DID parity (R side) -- Module 12.
#
# Reads data/12_sdid.csv (the StatsPAI California-Prop99 replica) and
# runs synthdid::synthdid_estimate. The headline ATT row is point-only;
# synthdid's native placebo SE is emitted as an explicitly named
# diagnostic row.

.args <- commandArgs(trailingOnly = FALSE)
.file_arg <- grep("^--file=", .args, value = TRUE)
.script_dir <- if (length(.file_arg) > 0) {
  dirname(normalizePath(sub("^--file=", "", .file_arg[1])))
} else {
  getwd()
}
source(file.path(.script_dir, "_common.R"))

suppressPackageStartupMessages({
  library(synthdid)
})

MODULE <- "12_sdid"

df <- read_csv_strict(MODULE)

# synthdid wants a panel matrix; build it via panel.matrices().
panel <- synthdid::panel.matrices(
  panel = df,
  unit = "state",
  time = "year",
  outcome = "cigsale",
  treatment = "treated"
)

set.seed(PARITY_SEED)
tau_hat <- synthdid::synthdid_estimate(
  Y = panel$Y, N0 = panel$N0, T0 = panel$T0
)

# Placebo SE: synthdid_se with method = "placebo".
se <- synthdid::synthdid_se(tau_hat, method = "placebo")

rows <- list(
  parity_row(
    module    = MODULE,
    statistic = "att_sdid",
    estimate  = as.numeric(tau_hat),
    se        = NA,
    ci_lo     = NA,
    ci_hi     = NA,
    n         = nrow(df)
  ),
  parity_row(
    module    = MODULE,
    statistic = "se_synthdid_placebo",
    estimate  = se,
    n         = nrow(df)
  )
)

write_results(MODULE, rows,
              extra = list(method = "synthdid_estimate",
                           N0 = panel$N0,
                           T0 = panel$T0,
                           se_method = "placebo",
                           se_reference = paste(
                             "R records synthdid_se placebo SEs as a",
                             "diagnostic row; att_sdid is point-only."
                           )))

File: tests/r_parity/12_sdid.py
"""StatsPAI Synthetic DID parity (Python side) -- Module 12.

Runs the **native Python** synthetic-DID estimator
(``sp.sdid(backend='native')``) on the california_prop99 replica
(Arkhangelsky et al. 2021). The companion 12_sdid.R uses
``synthdid::synthdid_estimate`` on the same CSV.

Tier: T2 native reference parity.  The native Python implementation now
mirrors synthdid's collapsed-form Frank-Wolfe weight solver and zeta
scaling, so the ATT matches the R reference on identical CSV bytes.  The
headline ``att_sdid`` row is point-only; backend-native placebo SEs are
reported as explicitly named diagnostic rows.
"""
from __future__ import annotations

import statspai as sp

from _common import PARITY_SEED, ParityRecord, dump_csv, write_results


MODULE = "12_sdid"


def main() -> None:
    # simulated=True is load-bearing: this module *writes* data/12_sdid.csv,
    # and the committed fixture (with the R and Stata goldens computed on
    # it) is the covariate-rich replica. The loader's default later became
    # the real ADH panel, so a bare california_prop99() call regenerates
    # the fixture from different bytes and the row stops being same-byte.
    df = sp.datasets.california_prop99(simulated=True)
    dump_csv(df, MODULE)

    fit = sp.sdid(
        df,
        outcome="cigsale",
        unit="state",
        time="year",
        treated_unit="California",
        treatment_time=1989,
        backend="native",
        seed=PARITY_SEED,
    )

    rows: list[ParityRecord] = [
        ParityRecord(
            module=MODULE, side="py", statistic="att_sdid",
            estimate=float(fit.estimate),
            se=None,
            ci_lo=None,
            ci_hi=None,
            n=int(len(df)),
        ),
        ParityRecord(
            module=MODULE, side="py", statistic="se_native_placebo",
            estimate=float(fit.se),
            se=None,
            ci_lo=None,
            ci_hi=None,
            n=int(len(df)),
        ),
    ]

    write_results(
        MODULE, "py", rows,
        extra={
            "method": "sdid",
            "n_treated": int(fit.model_info["n_treated"]),
            "n_control": int(fit.model_info["n_control"]),
            "T_pre": int(fit.model_info["T_pre"]),
            "T_post": int(fit.model_info["T_post"]),
            "se_method": fit.model_info["se_method"],
            "backend": fit.model_info.get("backend", "native"),
            "validation_tier": fit.model_info.get("validation_tier"),
            "reference_backend": fit.model_info.get("reference_backend"),
            "tier": "T2",
            "native_parity_note": (
                "Headline row is the NATIVE Python SDID ATT "
                "(backend='native'). The native solver mirrors synthdid's "
                "collapsed-form Frank-Wolfe weights and zeta scaling, so the "
                "ATT is a T2 same-byte reference-parity row. Backend-native "
                "placebo SEs are reported as separately named diagnostic rows."
            ),
            "se_reference": (
                "Python records StatsPAI deterministic all-control placebo "
                "SEs, R records synthdid_se placebo SEs, and Stata records "
                "sdid placebo SEs."
            ),
        },
    )


if __name__ == "__main__":
    main()
